File: java/recursions_backtrackings/Day10/Permutation_II.java
package recursions_backtrackings.Day10;

import java.util.*;

public class Permutation_II {
    private static void recurPermute_(int ind, int[] arr, List<List<Integer>> ans) {
        if (ind == arr.length) {
            List<Integer> ds = new ArrayList<>();

            for (int i = 0; i < arr.length; i++) {
                ds.add(arr[i]);
            }

            ans.add(new ArrayList<>(ds));
            return;
        }

        for (int i = ind; i < arr.length; i++) {
            swap(i, ind, arr);
            recurPermute_(ind + 1, arr, ans);
            swap(i, ind, arr);
        }
    }


    public static void swap(int i, int j, int[] arr) {
        int t = arr[i];
        arr[i] = arr[j];
        arr[j] = t;
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        System.out.print("Enter the size of arr: ");
        int n = sc.nextInt();
        int[] arr = new int[n];

        System.out.print("Enter " + n + " elements: ");
        
        for (int i = 0; i < n; i++) {
            arr[i] = sc.nextInt();
        }

        List<List<Integer>> ans = new ArrayList<>();
        recurPermute_(0, arr, ans);

        System.out.print("Permutations:");
        for (List<Integer> perm : ans) {
            System.out.println(perm);
        }

        sc.close();
    }
}

File: python/recursions_backtrackings/day10/permutation_II.py
def recurePermute_(ind, arr, ans):
    if ind == len(arr):
        ans.append(arr.copy())
        return
    
    for i in range(ind, len(arr)):
        swap(i, ind, arr)
        recurePermute_(ind + 1, arr, ans)
        swap(i, ind, arr)


def swap(i, j, arr):
    arr[i], arr[j] = arr[j], arr[i]



# -------- main --------
n = int(input("Enter the size of arr: "))
arr = list(map(int, input(f"Enter {n} elements: ").split()))

ans = []
recurePermute_(0, arr, ans)

print("Permutations:")
for perm in ans:
    print(perm)